frontend/deb/pkg.go

```diff
@@ -0,0 +1,128 @@
+package deb
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/Azure/dalec"
+	"github.com/moby/buildkit/client/llb"
+	"github.com/moby/buildkit/solver/pb"
+	"github.com/pkg/errors"
+)
+
+func mountSources(sources map[string]llb.State, dir string, mod func(string) string) llb.RunOption {
+	return dalec.RunOptFunc(func(ei *llb.ExecInfo) {
+		for key, src := range sources {
+			if mod != nil {
+				key = mod(key)
+			}
+			llb.AddMount(filepath.Join(dir, key), src).SetRunOption(ei)
+		}
+	})
+}
+
+func SourcePackage(sOpt dalec.SourceOpts, worker llb.State, spec *dalec.Spec, targetKey string, opts ...llb.ConstraintsOpt) (llb.State, error) {
+	dr, err := Debroot(sOpt, spec, llb.Scratch(), targetKey, "")
+	if err != nil {
+		return llb.Scratch(), err
+	}
+
+	sources, err := dalec.Sources(spec, sOpt)
+	if err != nil {
+		return llb.Scratch(), err
+	}
+
+	for _, patches := range spec.Patches {
+		for _, patch := range patches {
+			// patches do not need to be included in the sources
+			delete(sources, patch.Source)
+		}
+	}
+
+	patchesMountInput := llb.Scratch().
+		File(llb.Mkfile("dalec-changes.patch", 0o600, patchHeader))
+
+	// Our patches apply directly to each individual source tree, e.g. `cd <src>; patch ...`
+	// Debian applies patches from 1 directory up from the source tree (e.g. no `cd` as above).
+	// As such the patch files are not formatted correctly for Debian's build tooling.
+	// Here we generate a single patch file that generates the correct format.
+	//
+	// This way dpkg-source can automatically apply patches for us, and informs
+	// the caller of the patches applied and is generally just more inline with
+	// a typical deb build.
+	//
+	// This is using git instead of raw `diff` or other standalone tooling because only git appears to preserve permissions for new files.
+	// As an example, if patch adds a new file with mode +x, `diff` will not see the permissions for that new file.
+
+	patches := llb.Scratch()
+	if len(spec.Patches) > 0 {
+		patches = worker.
+			Run(dalec.ShArgs("set -e; git config --global user.email phony; git config --global user.name Dalec")).
+			Run(
+				dalec.ShArgs("set -e; git init .; git add .; git commit -m 'Initial commit'; \"${DEBIAN_DIR}/dalec/patch.sh\"; git add .; git commit -m 'With patch'; git diff HEAD~1 >> /work/out/dalec-changes.patch; echo 'dalec-changes.patch' > /work/out/series"),
+				llb.Dir("/work/sources"),
+				mountSources(sources, "/work/sources", nil),
+				// DEBIAN_DIR is used by the patch script to find the debian directory where we actually have the patches
+				llb.AddEnv("DEBIAN_DIR", "/work/debian"),
+				llb.AddMount("/work/debian", dr, llb.SourcePath("debian"), llb.Readonly),
+				dalec.WithConstraints(opts...),
+			).AddMount("/work/out", patchesMountInput)
+	}
```

Comment:
I think this is substantial and important enough to be moved into a separate function